[Site Isolation] Navigation to about:blank inherits navigator's origin

atar13 opened the pull request:
#### 84738405475ef105f150ad3d56be0e0c268902a2
<pre>
[Site Isolation] Navigation to about:blank inherits navigator&apos;s origin
<a href="https://bugs.webkit.org/show_bug.cgi?id=302102">https://bugs.webkit.org/show_bug.cgi?id=302102</a>
<a href="https://rdar.apple.com/160706600">rdar://160706600</a>

Reviewed by Alex Christensen.

If an iframe is navigated to about:blank, it should inherit the origin
of whoever initiated the navigation.

With site isolation, this means that the about:blank iframe needs to
be in the same process as the navigator. If the iframe was previously
in a different origin than the navigator, this might require a process
swap.

This patch also attempts to perform these loads to about:blank &quot;immediately&quot;
by committing a provisional frame right after creating one since about:blank
iframes don&apos;t have any network resources to load.

Tests: http/tests/site-isolation/navigate-to-about-blank-from-cross-origin-initiated-by-main-page.html
       http/tests/site-isolation/navigate-to-about-blank-from-cross-origin-initiated-by-parent-iframe.html

* LayoutTests/http/tests/site-isolation/navigate-to-about-blank-from-cross-origin-initiated-by-main-page-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/navigate-to-about-blank-from-cross-origin-initiated-by-main-page.html: Added.
* LayoutTests/http/tests/site-isolation/navigate-to-about-blank-from-cross-origin-initiated-by-parent-iframe-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/navigate-to-about-blank-from-cross-origin-initiated-by-parent-iframe.html: Added.
* LayoutTests/http/tests/site-isolation/resources/frame-child-navigate-to-about-blank-from-cross-origin-initiated-by-parent-iframe.html: Added.
* Source/WebKit/Shared/ProvisionalFrameCreationParameters.h:
* Source/WebKit/Shared/ProvisionalFrameCreationParameters.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
(WebKit::ProvisionalFrameProxy::ProvisionalFrameProxy):
* Source/WebKit/UIProcess/ProvisionalFrameProxy.h:
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::initializeWebPage):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::prepareForProvisionalLoadInProcess):
(WebKit::WebFrameProxy::provisionalFrameCreationParameters):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::continueNavigationInNewProcess):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigationInternal):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::createProvisionalFrame):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, MainPageNavigatesCrossOriginIframeToAboutBlank)):
(TestWebKitAPI::(SiteIsolation, ChildIframeNavigatesCrossOriginGrandchildIframeToAboutBlank)):

Canonical link: <a href="https://commits.webkit.org/303299@main">https://commits.webkit.org/303299@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c05ee53dd6f8281a26d2e7e442d31420aa16c0cb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131695 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4187 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42695 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139204 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83558 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b376962d-9c99-4a43-862f-0755f7d6dae7) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133565 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4113 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3949 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100674 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68111 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/43fd7279-dabd-4a0d-a9fb-3db401584a27) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134641 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2973 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117967 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81445 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/27a93207-a603-4feb-b5a5-f186a7cbb72f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2859 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/821 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82426 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111585 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36092 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141850 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3856 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36663 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109038 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3937 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3537 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109195 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2946 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114240 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/57066 "The change is no longer eligible for processing.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20518 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3909 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32658 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3741 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67357 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4001 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3870 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -474,6 +474,13 @@ void WebFrame::createProvisionalFrame(ProvisionalFrameCreationParameters&& param
+{
+    createProvisionalFrame(WTFMove(parameters));
+    commitProvisionalFrame();
+    send(Messages::WebFrameProxy::DidCreateFrameImmediately(m_frameID));
```
If this always sends a reply, can we use sendWithAsyncReply instead?  Also, the sender already has the frameID, so it doesn't need to be sent back.

atar13 replied:
Yeah an async reply is perfect here and removes the need for this new IPC message. Updated here: https://github.com/WebKit/WebKit/commit/efe8fbba17bbb1953a1619b801b74c984b532e08#diff-978336778d27e04722b8fecc60b0c6dcf8b42838c6b5c7bb80ebf6586f8a88c7R518-R524

achristensen07 commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -2367,6 +2373,7 @@ std::tuple<Ref<WebProcessProxy>, RefPtr<SuspendedPageProxy>, ASCIILiteral> WebPr
             return { WTFMove(sourceProcess), nullptr, "Navigation is treated as same-site"_s };
     }
 
+
```
nit: unnecessary newline

achristensen07 commented on Source/WebKit/UIProcess/WebFrameProxy.h:
```diff
@@ -313,6 +315,9 @@ class WebFrameProxy : public API::ObjectImpl<API::Object::Type::Frame>, public I
     WebCore::ScrollbarMode m_scrollingMode;
+
+    RefPtr<WebProcessProxy> m_processBeforeImmediateLoad;
+    RefPtr<FrameProcess> m_frameProcessAfterImmediateLoad;
```
Keeping a strong reference to a FrameProcess is probably not the right thing to do here.

atar13 replied:
I'm getting rid of `m_processBeforeImmediateLoad;` due to the change from this comment https://github.com/WebKit/WebKit/pull/52845/files#r2496519520.

However, I think I need `m_frameProcessAfterImmediateLoad` when the `didCreateFrameImmediately` function is called. So I might hold onto a `WeakPtr` and add null checks in `didCreateFrameImmediately`.

atar13 replied:
Since I moved to use `sendWithAsyncReply` (https://github.com/WebKit/WebKit/pull/52845#discussion_r2496516175), I no longer need to store the `RefPtr<FrameProcess>` as a member.

However, now I capture the `Ref` that `BrowsingContextGroup::ensureProcessForSite` returns. Maybe I should also only capture a `WeakPtr`.

This is what the code looks like now: https://github.com/WebKit/WebKit/commit/efe8fbba17bbb1953a1619b801b74c984b532e08#diff-978336778d27e04722b8fecc60b0c6dcf8b42838c6b5c7bb80ebf6586f8a88c7R516-R525
```cpp
Ref newFrameProcess = group.ensureProcessForSite(Site { effectiveOrigin }, mainFrameSite, newProcess, page->protectedPreferences(), InjectBrowsingContextIntoProcess::No);

newProcess.sendWithAsyncReply(Messages::WebFrame::CreateProvisionalFrameAndCommit(defaultProvisionalFrameCreationParameters()),
        [weakThis = WeakPtr { *this }, newFrameProcess] {
            RefPtr protectedThis = weakThis.get();
            if (!protectedThis)
                return;
            protectedThis->setProcess(newFrameProcess);
    }, frameID());
 ```

atar13 replied:
Changed to capture a `WeakPtr` to the `FrameProcess` https://github.com/WebKit/WebKit/compare/1b1f7df19bac2a32be564843e42a6d1dd1e25e67..516e581c8c1ce8c7c2b2309c24150a4696e5d0a2

achristensen07 commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -495,6 +495,50 @@ void WebFrameProxy::didCreateSubframe(WebCore::FrameIdentifier frameID, String&&
+
+    // Notify the old process that it's LocalFrame has been loaded in a different process
+    if (m_processBeforeImmediateLoad) {
+        m_processBeforeImmediateLoad->send(Messages::WebPage::LoadDidCommitInAnotherProcess(frameID, m_layerHostingContextIdentifier), protectedPage()->webPageIDInProcess(*m_processBeforeImmediateLoad));
```
This probably doesn't need to wait for the IPC roundtrip of CreateProvisionalFrameAndCommit to send.  Would it work to send it to the other process at the same time?

atar13 replied:
I think so. It has no dependency on the LocalFrame being created in the new process. I will give it a try in `createLocalFrameInProcessWithEffectiveOrigin`

atar13 replied:
Seems to have the same behavior, so updated here: https://github.com/WebKit/WebKit/commit/efe8fbba17bbb1953a1619b801b74c984b532e08#diff-978336778d27e04722b8fecc60b0c6dcf8b42838c6b5c7bb80ebf6586f8a88c7R512

achristensen07 commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -495,6 +495,50 @@ void WebFrameProxy::didCreateSubframe(WebCore::FrameIdentifier frameID, String&&
+        effectiveReferrerPolicy(),
+        scrollingMode(),
+        remoteFrameSize(),
+        }), frameID());
```
Indentation is weird here.  Also, could we make an abstraction on WebFrameProxy that returns a ProvisionalFrameCreationParameters and share it among the now 3 places it's used?

atar13 replied:
Yeah I'll make something like a `defaultProvisionalCreationParameters` which constructs these commonly used values.

atar13 replied:
Updated here https://github.com/WebKit/WebKit/commit/efe8fbba17bbb1953a1619b801b74c984b532e08#diff-978336778d27e04722b8fecc60b0c6dcf8b42838c6b5c7bb80ebf6586f8a88c7R880-R892. 

I only replaced it in my new case and in `ProvisionalFrameProxy.cpp`. There was another case in `ProvisionalPageProxy.cpp` but it was setting `frameIDBeforeProvisionalNavigation` differently.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5372,10 +5372,17 @@ void WebPageProxy::continueNavigationInNewProcess(API::Navigation& navigation, W
         loadParameters.isPerformingHTTPFallback = isPerformingHTTPFallback == IsPerformingHTTPFallback::Yes;
         loadParameters.isHandledByAboutSchemeHandler = m_aboutSchemeHandler->canHandleURL(loadParameters.request.url());
 
+
```
nit: newline

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5372,10 +5372,17 @@ void WebPageProxy::continueNavigationInNewProcess(API::Navigation& navigation, W
+        // for resources to load
+        bool shouldCommitImmediately = navigation.currentRequest().url().isAboutBlank();
+        // about:blank frames should be in the same process as the frame which originated navigation
+        std::optional<SecurityOriginData> originator = navigation.currentRequest().url().isAboutBlank() && navigation.originatingFrameInfo() ? std::make_optional(navigation.originatingFrameInfo()->securityOrigin) : std::nullopt;
```
We probably don't need to send both an optional and a bool.  If the optional isn't nullopt, that means we should commit it immediately

atar13 replied:
Updated to only have the optional https://github.com/WebKit/WebKit/compare/48c0cca62a1fc26f8ecc038a6df3c75b46f95bf9..aa4d4e998b288265a0006e89012d5a7269a3253e?diff=unified&w#diff-713e4ec79a64c4e41e2a80f757ce1074a599d2a2886c3cbaaba4ada4e531a625L5379-R5381

achristensen07 commented on Source/WebKit/UIProcess/ProvisionalFrameProxy.h:
```diff
@@ -41,7 +41,7 @@ class WebProcessProxy;
     WTF_MAKE_TZONE_ALLOCATED(ProvisionalFrameProxy);
 public:
-    explicit ProvisionalFrameProxy(WebFrameProxy&, Ref<FrameProcess>&&);
+    explicit ProvisionalFrameProxy(WebFrameProxy&, Ref<FrameProcess>&&, bool);
```
This would be more readable if you gave the bool a name or used a 2-state enum.

atar13 replied:
Yeah I agree this isn't great. It's already hard enough to understand unnamed function parameters in headers as it is and just a plain `bool` isn't very helpful.

Just changed to an `enum` in the latest commit https://github.com/WebKit/WebKit/compare/48c0cca62a1fc26f8ecc038a6df3c75b46f95bf9..aa4d4e998b288265a0006e89012d5a7269a3253e

achristensen07 commented on Source/WebKit/Shared/ProvisionalFrameCreationParameters.h:
```diff
@@ -38,6 +38,11 @@ using SandboxFlags = OptionSet<SandboxFlag>;
 
 namespace WebKit {
 
+enum class CommitTiming : uint8_t {
```
: bool

atar13 replied:
Updated in https://github.com/WebKit/WebKit/commit/805a9363457978b235d46a8c9297c0c361677e95#diff-7a605c863625769b80b480b4cd0c1d8979db619ccad714d4b1d153ee6dd69773R41-R44

achristensen07 commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -2270,6 +2270,11 @@ std::tuple<Ref<WebProcessProxy>, RefPtr<SuspendedPageProxy>, ASCIILiteral> WebPr
     }
 
+    if (siteIsolationEnabled && navigation.currentRequest().url().isAboutBlank()) {
+        RefPtr navigationOriginatorFrame = WebFrameProxy::webFrame(navigation.originatingFrameInfo()->frameID);
```
navigationOriginatorFrame and navigation.originatingFrameInfo should be null checked.

atar13 replied:
Added both null checks here https://github.com/WebKit/WebKit/commit/805a9363457978b235d46a8c9297c0c361677e95#diff-7625e12c845787a874aee6306cbd9a28fa43e425c69b6a45321aeba0f0e347d3R2275-R2278

achristensen07 commented on Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
```diff
@@ -39,22 +39,16 @@ namespace WebKit {
-        frame.remoteFrameSize()
-    }), frame.frameID());
+    auto parameters = frame.defaultProvisionalFrameCreationParameters();
+    parameters.commitTiming = commitTiming;
```
nit: I think it would be a slightly nicer design to pass commitTiming to the call to defaultProvisionalFrameCreationParameters as a parameter.  Also, I'd rename defaultProvisionalFrameCreationParameters to provisionalFrameCreationParameters

atar13 replied:
Updated to take commit timing here https://github.com/WebKit/WebKit/commit/805a9363457978b235d46a8c9297c0c361677e95#diff-978336778d27e04722b8fecc60b0c6dcf8b42838c6b5c7bb80ebf6586f8a88c7R879

I also made it take a `std::optional<WebCore::FrameIdentifier> frameIDBeforeProvisionalNavigation,` to replace another construction of `ProvisionalFrameCreationParameters` with this function https://github.com/WebKit/WebKit/commit/805a9363457978b235d46a8c9297c0c361677e95#diff-04ad8e0911b189ebd537422c9caf1fb20b8978506cda05d683f4f12845b03b95L298-R301

atar13 committed:
[Site Isolation] Navigation to about:blank inherits navigator's origin
https://bugs.webkit.org/show_bug.cgi?id=302102
rdar://160706600

Reviewed by Alex Christensen.

If an iframe is navigated to about:blank, it should inherit the origin
of whoever initiated the navigation.

With site isolation, this means that the about:blank iframe needs to
be in the same process as the navigator. If the iframe was previously
in a different origin than the navigator, this might require a process
swap.

This patch also attempts to perform these loads to about:blank "immediately"
by committing a provisional frame right after creating one since about:blank
iframes don't have any network resources to load.

Tests: http/tests/site-isolation/navigate-to-about-blank-from-cross-origin-initiated-by-main-page.html
       http/tests/site-isolation/navigate-to-about-blank-from-cross-origin-initiated-by-parent-iframe.html

* LayoutTests/http/tests/site-isolation/navigate-to-about-blank-from-cross-origin-initiated-by-main-page-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/navigate-to-about-blank-from-cross-origin-initiated-by-main-page.html: Added.
* LayoutTests/http/tests/site-isolation/navigate-to-about-blank-from-cross-origin-initiated-by-parent-iframe-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/navigate-to-about-blank-from-cross-origin-initiated-by-parent-iframe.html: Added.
* LayoutTests/http/tests/site-isolation/resources/frame-child-navigate-to-about-blank-from-cross-origin-initiated-by-parent-iframe.html: Added.
* Source/WebKit/Shared/ProvisionalFrameCreationParameters.h:
* Source/WebKit/Shared/ProvisionalFrameCreationParameters.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
(WebKit::ProvisionalFrameProxy::ProvisionalFrameProxy):
* Source/WebKit/UIProcess/ProvisionalFrameProxy.h:
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::initializeWebPage):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::prepareForProvisionalLoadInProcess):
(WebKit::WebFrameProxy::provisionalFrameCreationParameters):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::continueNavigationInNewProcess):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigationInternal):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::createProvisionalFrame):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, MainPageNavigatesCrossOriginIframeToAboutBlank)):
(TestWebKitAPI::(SiteIsolation, ChildIframeNavigatesCrossOriginGrandchildIframeToAboutBlank)):

Canonical link: https://commits.webkit.org/303299@main
